DEVELOPING.md

```diff
@@ -1,75 +1,210 @@
+Git's `rebase` command is much-feared but powerful, and you should become very familiar with it.
+If you are developing a feature in a branch, you can enter rebase mode by running:
+```bash
+git rebase -i master
+```
+This will take you to a text editor where you can command git to do all kinds of things with your commits!
+You can:
+ * Reorder them
+ * Squash several of them into a single commit
+ * Select them to be manually edited/amended
+
+Then, git will run through all your commits sequentially and execute whatever manipulations you indicated.
+Squashing is very useful for combining those periodic check-in commits into something meaningful.
+Editing is useful for responding to PR review comments; most people know about using `git --amend` to apply changes to the latest commit, but using the `rebase` edit option lets you apply changes to commits in the past!
 
-The `tlatools.jar` will be created in the `org.lamport.tlatools/dist` directory.
+### Splitting up commits and PRs
 
-Test that the build was successful by using an example model:
+While rebasing, you can also use editing to split a single commit up into multiple commits.
+This is fairly straightforward when the changes are in different files, but if you want some changes within a single file to go in one commit and some to go in another commit, you can accomplish this interactively with:
+```bash
+git add --patch <filename>
+```
+Sometimes you might want to split your commits across multiple branches so they can be in separate PRs.
+For this, switch to a new branch then use:
 ```bash
-$ java -jar dist/tla2tools.jar ./test-model/pcal/Bakery.tla
+git cherry-pick <commit hash>
 ```
-To run the unit tests, abbreviate the above `ant` command:
+This will add the specified commit to the head of your new branch.
+You can then rebase your original feature branch to remove that commit.
+
+### Developer Certificate of Origin (DCO) signing
+
+Due to legal disputes in the mid-2000s, all projects under the Linux Foundation are required to have commits be signed with a [Developer Certificate of Origin (DCO)](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin).
+This basically says that you (the developer) were entirely responsible for writing the code in the commit and that you are legally permitted to contribute it under a permissive license (it isn't copied from proprietary code, for example).
+In git this is easily done by adding an extra `-s` flag as you commit:
 ```bash
-$ ant -f customBuild.xml test
+git commit -s -am "Commit message here"
 ```
+Don't worry too much about forgetting this; the CI will catch it and GitHub provides a page with simple instructions to retroactively add DCO signing to past commits.
+Eventually adding the `-s` flag will become muscle memory.
 
-If you're using Linux or OSX, you can install `entr`,
-then run the following to watch for and automatically recompile changes:
+Other Useful Tools
```

Comment:
This overhaul of developing seems like a good opportunity to get rid of promoting entr.